Migrate from cloud > self-hosted (#5206)
mattermost/docs: source/manage/cloud-data-export.rst

Cloud Data Export
=========================
Mattermost Workspace Migration
==============================

> If, after completing your 14-day free Cloud trial, you've decided to self-host your own Mattermost deployment, you may want to migrate your Cloud data to your self-hosted instance.

Reviewer: I wonder if the phrase "Mattermost instance" is better here (and I see this is used elsewhere)

> This document outlines the process for migrating from Mattermost Cloud to a Mattermost self-hosted deployment. In the future, a process for migrating from Mattermost self-hosted to Mattermost Cloud will also be documented and provided here.

Author: ```suggestion
This document outlines the process for migrating from Mattermost Cloud to a Mattermost self-hosted instance. In the future, a process for migrating from Mattermost self-hosted to Mattermost Cloud will also be documented and provided here.
```

> Migrating between two installations is the same process regardless as to whether the source or destination of the migration is in the Cloud or self-hosted. **These steps will work for any Mattermost instance**, as long as the instances involved have a version of Mattermost later than 5.33.

Reviewer 2: > Migrating between two installations is the same process regardless as to whether the source or destination of the migration is in the Cloud or self-hosted

It's not clear what is meant by `the same process` in this context?

Author: Thanks for catching this - it's from Ian's blog post so it references a bit that I took out.

> Migrating between two installations is the same process regardless as to whether the source or destination of the migration is in the Cloud or self-hosted. **These steps will work for any Mattermost instance**, as long as the instances involved have a version of Mattermost later than 5.33.

Author: ```suggestion
Migrating between two installations follows the same process that's documented below, regardless as to whether the source or destination of the migration is in the Cloud or self-hosted. **These steps will work for any Mattermost instance**, as long as the instances are on Mattermost version 5.33 and up.
```

Migrating from Cloud to self-hosted
-----------------------------------

If, after completing your 14-day free Cloud trial, you've decided to self-host your own Mattermost deployment, you can migrate your Cloud data to your self-hosted instance.

How does the process work?
--------------------------

> Before you export and migrate your data, you must `install Mattermost <https://docs.mattermost.com/guides/deployment.html#install-guides>`_ on the server you’ll be using to run Mattermost. The migration is done using the mmctl CLI tool, which is a remote CLI tool for Mattermost that's installed locally and uses the Mattermost API. From Mattermost version 6.0 onwards, mmctl is pre-installed. For versions prior to 6.0, first download the newest release from [the releases page](https://github.com/mattermost/mmctl/releases/tag/v6.1.0), `install mmctl <https://docs.mattermost.com/manage/mmctl-command-line-tool.html#install-mmctl>`_, and place the binary on your ``PATH``.

Reviewer 2: ```suggestion
Before you export and migrate your data, you must `install Mattermost <https://docs.mattermost.com/guides/deployment.html#install-guides>`_ on the server you’ll be using to run Mattermost. The migration is done using the mmctl CLI tool, which is a remote CLI tool for Mattermost that's installed locally and uses the Mattermost API. From Mattermost version 6.0 onwards, mmctl is pre-installed. For versions prior to 6.0, first download the newest release from `the releases page <https://github.com/mattermost/mmctl/releases/tag/v6.1.0>`_, `install mmctl <https://docs.mattermost.com/manage/mmctl-command-line-tool.html#install-mmctl>`_, and place the binary on your ``PATH``.
```

> Before you export and migrate your data, you must `install Mattermost <https://docs.mattermost.com/guides/deployment.html#install-guides>`_ on the server you’ll be using to run Mattermost. The migration is done using the mmctl CLI tool, which is a remote CLI tool for Mattermost that's installed locally and uses the Mattermost API. From Mattermost version 6.0 onwards, mmctl is pre-installed. For versions prior to 6.0, first download the newest release from `the releases page <https://github.com/mattermost/mmctl/releases/tag/v6.1.0>`_, `install mmctl <https://docs.mattermost.com/manage/mmctl-command-line-tool.html#install-mmctl>`_, and place the binary on your ``PATH``.

Reviewer 3: ```suggestion
Before you export and migrate your data, you must `install Mattermost <https://docs.mattermost.com/guides/deployment.html#install-guides>`_ on the server you’ll be using to run Mattermost. The migration is done using the mmctl CLI tool, which is a remote CLI tool for Mattermost that's installed locally and uses the Mattermost API. From Mattermost v6.0 onwards, mmctl is pre-installed. For versions prior to 6.0, first download the newest release from `the releases page <https://github.com/mattermost/mmctl/releases/tag/v6.1.0>`_, `install mmctl <https://docs.mattermost.com/manage/mmctl-command-line-tool.html#install-mmctl>`_, and place the binary on your ``PATH``.
```

Reviewer 3: I've created [a ticket on our v6.2 board](https://community-daily.mattermost.com/boards/workspace/g4k9s1f953n4bytz5f9sz9orer/bm1j7wcrrspoymkjermbydid97h/vcwyy7xnhh4qits4thrb3pte53e/cc3j1mtw7ytyhuckm5tceigwdqh) to update the mmctl docs to clarify that mmctl comes pre-installed as of v6.0 and later.

You'll be using the `mmctl export <https://docs.mattermost.com/manage/mmctl-command-line-tool.html#mmctl-export>`__ commands to export your Cloud data for channels, messages, users, etc. The export file is downloaded to a location specified in the export commands. Once the export is complete, you'll import the data into your self-hosted instance.

.. note::
  
  The export process doesn’t include integrations or any custom data. Other aspects of your instance, such as specific security settings and requirements, are also not included. For assistance with migrating additional data and settings, see our support options: https://mattermost.com/support/.

Once mmctl is installed, we can generate the export from the source instance.

Authenticate
------------

Authentication is done with either Mattermost login credentials or an authentication token. First, use your administrator credentials to log into the instance with mmctl, replacing `example-source-domain.com` with the network address of the source instance:

.. code::

>    $ mmctl auth login https://example-source-domain.com

Reviewer: ```suggestion
   $ mmctl auth login https://yourdomain.cloud.mattermost.com
```

.. code::

  $ mmctl export job show yfrr9ku5i7fjubeshs1ksrknzc
    ID: yfrr9ku5i7fjubeshs1ksrknzc
    Status: success
    Created: 2021-11-03 10:44:13 -0500 CDT
    Started: 2021-11-03 10:44:23 -0500 CDT

Download the export
-------------------

Once the status is ``success``, you can download the export onto your local machine. First, discover the name of the completed export file with ``mmctl export list``:

.. code::

   $ mmctl export list
   r3kcj8yuwbramdt714doafi3oo_export.zip

This will show all of the exports on the server, so be sure to download the latest one and to delete it when you're done to save storage. Download the file with a command like the following, but with the filename of the export on your server:

.. code::

   $ mmctl export download r3kcj8yuwbramdt714doafi3oo_export.zip

Upload the export to the new server
-----------------------------------

Finally, it's time to take our export from the source server and use it as an import into the destination server. First, log into the destination server using mmctl the same way you logged into the source server:

.. code::

>    $ mmctl auth login https://my-destintation-server.example

Reviewer: ```suggestion
   $ mmctl auth login https://localinstance.company.com
```

Use the following command to upload the export to the destination server. The speed of the upload may vary based on connection speed. When the upload is complete the command will return with the ID of the import:

.. code::
  

>    mmctl import upload r3kcj8yuwbramdt714doafi3oo_export.zip

Reviewer 4: One thing to note that I discovered after writing my blog post is that this step is subject to the upload file size limit specified in the System Console. On Cloud, and by default I believe, this is 100MB. For self-hosted -> Cloud, if the import is >100MB in size, the user will get stuck and have to contact support at this step. For Cloud -> Self-Hosted, the user may hit a bump here and need to go into the System Console to raise their upload limit to something larger than the attachment upload.

I'm 0/5 on this but long term I think it would be worth investigating whether or not uploaded archives being sent via `mmctl` should be subject to the same file upload limit as files that are uploaded into Channels via the web interface.

Reviewer 3: Great point, @gigawhitlocks! @justinegeffen @nab-77 for visibility

Author: Thanks for bringing this up. I did see your comment about this but it slipped my mind. 

I think one way of circumventing this is to recommend changing the limit prior to the process of cloud > self-hosted. Do we know the average size of the upload? 

Assuming it's not possible to change this setting for Cloud, what happens to the user's process? Do they need to restart?

Reviewer 4: For imports to Cloud, they'll have to reach out to support and have the process handled by us. Support can upload the archive to the AWAT which will import it in a way that is not subject to this limitation. Their workspace won't be in a bad state or anything, it'll just be empty. Upload will fail, is all.

>one way of circumventing this is to recommend changing the limit prior to the process of cloud > self-hosted. Do we know the average size of the upload?

The only way to circumvent it is to change the limit prior to the upload. The size of the upload is entirely dependent upon the size of the source installation, so there's no way to estimate what an average size might be without collecting a lot of data from on-premise installations. After the export is generated, the user should look at the archive in a file manager (like Windows Explorer or macOS Finder) to see how big it is, and raise the limit accordingly, before attempting to upload it to the destination.

Log into your Cloud instance and run the following mmctl command: 

.. code:: none

   mmctl export create --attachments [flags]. 

Ensure you set **attachments** to **true** to include file attachments - this creates a zip file of your Mattermost data.

Next, download the export file by running the following mmctl command:

.. code::

   mmctl export download [exportname] [filepath] [flags]

.. note::

  You can indicate the name of the export and its destination path ``$ mmctl export download samplename sample_export.zip``. If you only indicate the name, the path
  will match: ``$ mmctl export download samplename``.

When the file download is complete, log into your Mattermost server. Run the following mmctl command to import your data to your Mattermost server:

.. code::
  
   mmctl import upload [filepath] [flags]. 
   
Then extract the file to use it by running the following mmctl command:

.. code::
   
   mmctl extract run [flags].

> Once your migration is complete and you’ve imported your data into your self-hosted instance. We recommend that you take a few days to validate your data and ensure everything is working as expected before taking down your Cloud instance.

Author: ```suggestion
Once your migration is complete and you’ve imported your data into your self-hosted instance we recommend that you take a few days to validate your data and ensure everything is working as expected before taking down your Cloud instance.
```

Once you've installed mmctl, you'll be using the `mmctl export <https://docs.mattermost.com/manage/mmctl-command-line-tool.html#mmctl-export>`__ commands to export your Cloud data for channels, messages, users, etc. The export file is downloaded to a location specified in the export commands. Once the export is complete, you'll import the data into your self-hosted instance.

.. note::
  
  The export process doesn’t include integrations or any custom data. Other aspects of your instance, such as specific security settings and requirements, are also not included. For assistance with migrating additional data and settings, see our support options: https://mattermost.com/support/.

Log into your Cloud instance and run the following mmctl command: 

.. code:: none

   mmctl export create --attachments [flags]. 

Ensure you set **attachments** to **true** to include file attachments - this creates a zip file of your Mattermost data.

> 

Author: ```suggestion
You can see the progress on the export job with ``mmctl export job list`` and when it's finished, the list of exported zipfiles with ``mmctl export list`` and from it get the exportname to use in the next command.
```

Changing Deployment Types
==========================

> This document outlines the process for migrating from Mattermost Cloud to Mattermost self-hosted. In the future, a process for migrating from Mattermost self-hosted to Mattermost Cloud will also be documented and provided here.

Author: ```suggestion
The process for migrating from Mattermost Cloud to Mattermost self-hosted. In the future, a process for migrating from Mattermost self-hosted to Mattermost Cloud will also be documented and provided here.
```
